### PR TITLE
Remove explicit capture of fields

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Previously we were capturing directly fields of the current class in a specific way and use them for condition evaluation, while this was captured as argument. Now we only capture this and all evaluations are based on implicit this dereferencing when needed. NB: Serialization of complex classes no longer capture static fields

# Motivation
Remove useless captures & performance

# Additional Notes

Jira ticket: [DEBUG-2399]

[DEBUG-2399]: https://datadoghq.atlassian.net/browse/DEBUG-2399?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/CapturedContext.java [hunk 1]
```diff
@@ -377,8 +357,8 @@ public String toString() {
         + locals
         + ", throwable="
         + throwable
-        + ", fields="
-        + fields
+        + ", staticFields="
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/CapturedContext.java:360]: good catch

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/RefResolverHelper.java [hunk 2]
```diff
@@ -3,12 +3,12 @@
 import datadog.trace.bootstrap.debugger.CapturedContext;
 import datadog.trace.bootstrap.debugger.el.ValueReferenceResolver;
 import datadog.trace.bootstrap.debugger.util.Redaction;
-import java.lang.reflect.Field;
 import java.util.*;
 
 public class RefResolverHelper {
 
   public static ValueReferenceResolver createResolver(Object instance) {
+    /*
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/RefResolverHelper.java:11]: remove comment?

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/ProbeConditionTest.java [hunk 3]
```diff
@@ -207,56 +207,70 @@ void stringPrimitives() throws IOException {
...
+        strSet.add("foo");
+      }
+
+      List<String> strList = new ArrayList<>();
+
+      {
+        strList.add("foo");
+      }
+    }
+    ValueReferenceResolver ctx =
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/ProbeConditionTest.java:272]: why not use createResolver(new Object)? - as it seems to do the same thing... 
```
public static ValueReferenceResolver createResolver(Object instance) { ... }
```

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 4]
```diff
@@ -857,7 +857,7 @@ private void collectStaticFields(InsnList insnList) {
     // stack: [capturedcontext]
   }
 
-  private void collectFields(InsnList insnList) {
+  private void collectSpecialFields(InsnList insnList) {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:860]: nit: rename to collectCorrleationInfo?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 5]
```diff
@@ -869,96 +869,27 @@ private void collectFields(InsnList insnList) {
...
+    // stack: [capturedcontext]
+    extractSpecialId(insnList, "dd.span_id", "getSpanId", "addSpanId");
+    // stack: [capturedcontext]
+  }
+
+  private void extractSpecialId(
+      InsnList insnList, String fieldName, String getMethodName, String addMethodName) {
     insnList.add(new InsnNode(Opcodes.DUP));
     // stack: [capturedcontext, capturedcontext]
-    ldc(insnList, fieldsToCapture.size());
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:880]: nice!!!

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ValueRefExpressionTest.java [hunk 6]
```diff
@@ -48,31 +49,23 @@ void testPredicatedRef() {
 
     RuntimeException runtimeException =
         assertThrows(RuntimeException.class, () -> isEmptyInvalid.evaluate(ctx));
-    assertEquals("Cannot find symbol: x", runtimeException.getMessage());
+    assertEquals("Cannot dereference to field: x", runtimeException.getMessage());
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ValueRefExpressionTest.java:52]: not sure `Cannot dereference to field` is right... 

I think `Cannot dereference field:` is better

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java [hunk 7]
```diff
@@ -189,11 +194,11 @@ public CapturedContext fromJson(JsonReader jsonReader) throws IOException {
             List<CapturedContext.CapturedValue> argValues = new ArrayList<>();
             while (jsonReader.hasNext()) {
               String argName = jsonReader.peekJson().nextName();
-              if ("this".equals(argName)) {
-                jsonReader.nextName(); // consume "this"
-                fromJsonFields(jsonReader, capturedContext);
-                continue;
-              }
+              //              if ("this".equals(argName)) {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java:197]: remove comment

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java [hunk 8]
```diff
@@ -234,7 +239,7 @@ private void fromJsonFields(JsonReader jsonReader, CapturedContext capturedConte
             }
           case FIELDS:
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java:240]: do we need this anymore?

**Review by shatzi — APPROVED** [review]

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/RefResolverHelper.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/RefResolverHelper.java]: done

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/ProbeConditionTest.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/ProbeConditionTest.java]: - true!

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ValueRefExpressionTest.java (hunk 6 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ValueRefExpressionTest.java]: done

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java (hunk 7 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java]: done

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java (hunk 8 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/util/MoshiSnapshotTestHelper.java]: no, removed

**Pull request merged**